### PR TITLE
Implement Reflect for TypeId

### PR DESCRIPTION
**Pull request by PROMETHIA-27**
# Objective

Allows reflection and serialization of TypeIds. This is helpful for serializing and deserializing certain types that use TypeIds and are otherwise reflectable/serializable. 

## Solution

- Implement TypeIds as reflect values with Hash and PartialEq
- Hardcode TypeId serialization as an alternative to Value serialization, since TypeIds need access to the TypeRegistry to be serialized/deserialized and otherwise cannot access it. TypeIds are de/serialized as their corresponding type's name.

## Issues:

- Hardcoding a new serialization method for one type is gross, especially when it's entirely possible other types might end up also needing a TypeRegistry for serialization.
- Serialization format can be simplified to `TypeId("{type name here}")` once [4561](https://github.com/bevyengine/bevy/pull/4561) hits

## Alternative solutions:

- Don't implement reflect for TypeIds
   - Would entirely prevent behavior which is probably logically sound and useful
- Implement reflect without serialization
   - It should be sound to allow TypeIds to be serialized to a type name and back, so this doesn't seem like a good approach
- Allow any serialized types to access the TypeRegistry during de/serialization
   - Add it to thread local storage?
   - Pass it as a parameter?

### CODE REVIEW

File: crates/bevy_reflect/src/serde/ser.rs [hunk 1]
```diff
@@ -107,6 +115,31 @@ impl<'a> Serialize for ReflectValueSerializer<'a> {
     }
 }
 
+pub struct TypeIdSerializer<'a> {
```
[the remark this answers is not in the document]
Comment (PROMETHIA-27) [reply crates/bevy_reflect/src/serde/ser.rs:118]: Yes, I wouldn't serialize TypeIds directly. I'm also aware type names are unstable, but if that instability manifests then all of bevy's reflection serialization has problems. I believe #4561 mentioned above will mitigate this somewhat but it'll still be an issue. Hopefully rust gets stable type names at some point?

**Review by MrGVSV — APPROVED** [review]
Looks good to me!

**Comment by MrGVSV** [comment]
> Allow any serialized types to access the TypeRegistry during de/serialization
> * Add it to thread local storage?
> * Pass it as a parameter?

Reflecting on this more, I think you're right. It's probably best to allow the serialization method to access the registry (even if #4782 lands) as this gives the best control and flexibility for types going forward.

We could probably add that functionality to #4782 itself or create a new PR for it (I don't think it should be included as part of this one).

**Comment by PROMETHIA-27** [comment]
Agreed, that's a big change beyond the scope of this PR. I think it probably belongs in a new PR. I'm also not certain on how it should be done.

**Comment by MrGVSV** [comment]
> I'm also not certain on how it should be done.

I’m in favor of the parameter method. The TLS solution will probably be a lot more controversial and a bigger change. 

So just adding it as a parameter is probably simplest and most likely to be approved for now

**Comment by PROMETHIA-27** [comment]
I'll investigate that and open a PR if I can get it working

**Review by alice-i-cecile — CHANGES_REQUESTED** [review]
@PROMETHIA-27, once this is rebased and the serialization fixed (remember to fix the PR description) I'm happy to merge this.

Other code LGTM.

**Comment by PROMETHIA-27** [comment]
Making a new PR for this soon:tm:

**Pull request closed without merging**